Rework Darwin BLE layer

ksperling-apple opened the pull request:
- Use the CHIP queue instead of a separate BLE queue
- Retain the underlying CBPeripheral in MTRCommissionableBrowserResult

### Testing

New automated tests based on mocking CoreBluetooth.
Manual testing with Mac and iOS pairing an M5Stack.

ksperling-apple committed:
Rename Ble*Delegate headers to match .mm / class names

ksperling-apple committed:
Simplify BLEManagerImpl, we don't need to allocate the DelegateImpls separately

ksperling-apple committed:
Add some tests that exercise BLE code by mocking Core Bluetooth

Build with Bluetooth enabled in CI and enable Bluetooth mocking for all tests.

ksperling-apple committed:
Retain the underlying CBPeripheral in MTRCommissionableBrowserResult

This avoids a dangling pointer if the BleConnectionDelegateImpl drops the
peripheral from the cache while a client still holds the browser result.

Also document what BLE_CONNECTION_OBJECT is on Darwin.

ksperling-apple committed:
Use the CHIP queue instead of a separate BLE queue

ksperling-apple committed:
Tidy up BLE delegate headers

ksperling-apple committed:
Factor out connObj bridge casts into helper functions

ksperling-apple committed:
Simplify UUID handling, CBUUID does most of it already

bzbarsky-apple commented on src/platform/Darwin/BleUtils.mm:
```diff
@@ -0,0 +1,65 @@
+            // CBUUID handles the expansion to 128 bit automatically internally,
+            // but doesn't expose the expanded UUID in the `data` property, so
+            // we have to re-implement the expansion here.
+            constexpr ChipBleUUID baseUuid = { { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x10, 0x00, 0x80, 0x00, 0x00, 0x80, 0x5F, 0x9B, 0x34, 0xFB } };
```
Might be worth documenting where the magic bytes come from?

ksperling-apple commented on src/platform/Darwin/BleConnectionDelegateImpl.mm:
```diff
@@ -613,20 +560,19 @@ - (void)stop
     // the centralManager and the peripheral members too early it won't be
     // able to reach those.
     // This is why closing connections happens as 2 async steps.
-    dispatch_async(_chipWorkQueue, ^{
```
Hm I thought I had updated that comment already... Yes `stop` is only called from the Matter queue. Everything in this file is now happening on the Matter queue.

ksperling-apple replied:
This is another internal helper that's only called from places already on the Matter queue, so I've left the assert off this one.

ksperling-apple committed:
Review comments / tidyup

Add assertChipStackLockedByCurrentThread() to CB delegate methods.
Use assign instead of unsafe_unretained for non-retainable types.
Add comment to BLE Base UUID.

ksperling-apple committed:
Avoid leaks false positive in CoreBluetooth mock

ksperling-apple commented on src/platform/Darwin/BleConnectionDelegateImpl.mm:
```diff
@@ -318,20 +292,16 @@ - (void)clearTimer
-            self.onConnectionError(self.appState, error);
-        }
-    });
+    if (self.onConnectionError != nil) {
```
It was already on the ConnectionDelegateImpl entry points (NewConnection etc) and I've added it to all the CBCentralManager / CBPeripheral delegate methods. With those all in place I think it's fine to leave them off any internal helper methods, since literally everything in this file now runs on the Matter queue.

ksperling-apple commented on src/platform/Darwin/BleConnectionDelegateImpl.mm:
```diff
@@ -318,20 +292,16 @@ - (void)clearTimer
-            self.onConnectionComplete(self.appState, (__bridge void *) peripheral);
-        }
-    });
+    if (self.onConnectionComplete != nil) {
```
As above

ksperling-apple commented on src/platform/Darwin/BleConnectionDelegateImpl.mm:
```diff
@@ -680,9 +626,7 @@ - (void)detachScannerDelegate
     auto * existingDelegate = _scannerDelegate;
```
This is another internal helper that's only called from places already on the Matter queue, so I've left the assert off this one.

ksperling-apple commented on src/platform/Darwin/BleConnectionDelegateImpl.mm:
```diff
@@ -693,11 +637,9 @@ - (void)updateWithDelegate:(chip::DeviceLayer::BleScannerDelegate *)delegate pre
     if (delegate) {
```
as above

ksperling-apple commented on src/platform/Darwin/BleConnectionDelegateImpl.mm:
```diff
@@ -768,12 +710,10 @@ - (void)addPeripheralToCache:(CBPeripheral *)peripheral data:(NSData *)data
         ChipLogProgress(Ble, "Adding peripheral %p to the cache", peripheral);
```
as above

ksperling-apple commented on src/platform/Darwin/BleConnectionDelegateImpl.mm:
```diff
@@ -820,9 +760,7 @@ - (void)removePeripheralFromCache:(CBPeripheral *)peripheral
 
         auto delegate = _scannerDelegate;
         if (delegate) {
-            dispatch_async(_chipWorkQueue, ^{
```
as above

bzbarsky-apple commented on src/platform/Darwin/BlePeripheral.h:
```diff
@@ -17,10 +17,29 @@
+namespace DeviceLayer {
+namespace Internal {
+
+// Holds a CBPeripheral and the associated CBCentralManager
+struct BlePeripheral
```
This struct seems t o be completely unused?

ksperling-apple replied:
Ah yes that's a remnant of my attempt to use a shared_ptr as the BLE_CONN_OBJ type, will remove for now.

ksperling-apple replied:
Hm the whole header is actually unused, I think it slipped into a commit by accident.

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRBleTests.m:
```diff
@@ -0,0 +1,192 @@
+{
+    [sController shutdown];
+    sController = nil;
+    [[MTRDeviceControllerFactory sharedInstance] stopControllerFactory];
```
Followup: MTRTestCase should perhaps handle this, since it handled the factory startup....  Need to think about it (which is why it's a followup).

ksperling-apple replied:
Yeah we would need some way to indicate if it's meant to be test scope or suite scope... maybe a class vs instance factory method, and then cleanup happens automatically with the correct scope.

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRBleTests.m:
```diff
@@ -0,0 +1,192 @@
+    XCTAssertTrue([sController stopBrowseForCommissionables]);
+
+    // Attempt to use the MTRCommissionableBrowserResult after we stopped browsing
+    // TODO: This currently results in a UAF because BLE_CONNECTION_OBJECT is a void*
```
Is this TODO comment still relevant?

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRBleTests.m:
```diff
@@ -0,0 +1,192 @@
+    XCTAssertTrue([sController setupCommissioningSessionWithPayload:payload newNodeID:@999 error:&error],
+        "setupCommissioningSessionWithPayload failed: %@", error);
+
+    // Create a race between shutdown and a CBManager callback that provokes a UAF
```
That's before the fixes in this PR, right?  So "used to provoke a UAF"?

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/TestHelpers/MTRMockCB.m:
```diff
@@ -0,0 +1,690 @@
+
+- (instancetype)_initWithDetails:(MTRMockCBPeripheralDetails *)details manager:(MTRMockCBCentralManager *)manager;
+
+@property (readonly, strong, nonatomic) MTRMockCBCentralManager * manager; // not API, but used by BlePlatformDelegateImpl via KVC
```
Not sure what KVC is here...

ksperling-apple replied:
Key Value Coding, it's another of those ObjC duck typing / reflection things... Our BLE Platform Manager does this to access this internal property:
```
CBCentralManager * manager = (CBCentralManager *) [peripheral valueForKey:@"manager"];
```

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/TestHelpers/MTRTestCase.mm:
```diff
@@ -106,6 +122,22 @@ - (void)tearDown
+    __auto_type * testKeys = [[MTRTestKeys alloc] init];
+    __auto_type * params = [[MTRDeviceControllerStartupParams alloc] initWithIPK:testKeys.ipk fabricID:@1 nocSigner:testKeys];
+    params.vendorID = @0xFFF1;
+    MTRDeviceController * controller = [factory createControllerOnNewFabric:params error:nil];
```
I know you copied this, but followup: we should probably change this to use the new controller-creation APIs.

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/TestHelpers/MTRMockCB.m:
I have to admit I did not check every single detail here carefully....

ksperling-apple committed:
Final round of review comments

ksperling-apple commented:
Fast-tracking Darwin-only change
